File: CppNaNa/change.cpp
// Welcome to qq group: 1030115250
#include "change.hpp"
#include <ctime>

bool ChangeManager::changepassword(const std::string& username,
    const std::string& newpassword,
    const std::string& question,
    const std::string& answer) {
    try {
        // 发送初始加密请求
        json initResponse = sendEncryptRequest();

        // 解密响应数据
        std::string decrypted = AES256GCM::decryptData(
            initResponse["data"].get<std::string>(),
            keyBuffer,
            initResponse["iv"].get<std::string>(),
            initResponse["authTag"].get<std::string>()
        );
        json decryptedData = json::parse(decrypted);
        netdata.aeskey = AES256GCM::base64_decode(decryptedData["data"].get<std::string>());

        // 构建请求数据
        json data = {
            {"name", username},
            {"newpassword", sha512(newpassword)},
            {"question", sha512(question)},
            {"answer", sha512(answer)},
            {"timestamp", LoginManager::timestamp_millis()}
        };

        // 加密数据
        EncryptedData encryptedData = AES256GCM::encryptJSON(data.dump(), netdata.aeskey);

        // 构造请求体
        json requestBody = {
            {"data", encryptedData.data},
            {"tag", encryptedData.tag},
            {"iv", encryptedData.iv}
        };

        std::string postData = requestBody.dump();
        // 发送修改密码请求
        std::string readBuffer;
        curl_easy_setopt(curl, CURLOPT_URL, (hostaddr + "/api/v1/users/change").c_str());
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postData.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, postData.size());

        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);

        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Content-Type: application/json");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        CURLcode res = curl_easy_perform(curl);
        curl_slist_free_all(headers);

        if (res != CURLE_OK) {
            throw std::runtime_error("Change password request failed: " +
                std::string(curl_easy_strerror(res)));
        }

        // 处理响应
        long response_code;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
        if (response_code != 200) {
            json errorData = json::parse(readBuffer);
            throw std::runtime_error("Change password failed: " +
                errorData["error"].get<std::string>() +
                " (原因: " + errorData["reason"].get<std::string>() + ")");
        }

        // 解密响应数据
        json responseData = json::parse(readBuffer);
        std::string deDataStr = AES256GCM::decryptData(
            responseData["data"].get<std::string>(),
            netdata.aeskey,
            responseData["iv"].get<std::string>(),
            responseData["tag"].get<std::string>()
        );

        json deData = json::parse(deDataStr);
        if (deData.contains("newkey")) {
            netdata.aeskey = AES256GCM::base64_decode(deData["newkey"].get<std::string>());
        }

        return deData["success"].get<bool>();
    }
    catch (const std::exception& e) {
        throw std::runtime_error("Password change failed: " + std::string(e.what()));
    }
}

File: x86lib/src/network/change.rs
// change.rs
use reqwest;
use serde_json::{Value, json};
use base64::{Engine as _, engine::general_purpose};
use openssl::hash::{hash, MessageDigest};
use crate::network::aesnet::{decrypt_data, encrypt_json};
use crate::network::host::{get_host, NETDATA, KEY_BUFFER};
use crate::network::login::send_encrypt_request;
use anyhow::{Context, Result};

pub async fn changepassword(
    username: &str,
    newpassword: &str,
    question: &str,
    answer: &str,
) -> Result<bool> {
    // 发送初始化请求
    let init_data = send_encrypt_request().await?;
    let ciphertext = init_data["data"].as_str().context("无效响应格式")?;
    let iv = init_data["iv"].as_str().context("无效响应格式")?;
    let auth_tag = init_data["authTag"].as_str().context("无效响应格式")?;

    // 解密并更新 AES 密钥
    let decrypted = decrypt_data(ciphertext, &KEY_BUFFER, iv, auth_tag)
        .map_err(|e| anyhow::anyhow!("初始解密失败: {}", e))?;

    *NETDATA.aeskey.lock().unwrap() = general_purpose::STANDARD
        .decode(decrypted["data"].as_str().context("无效密钥格式")?)?;

    // 生成哈希
    let newpassword_hash = hex::encode(hash(MessageDigest::sha512(), newpassword.as_bytes())?);
    let question_hash = hex::encode(hash(MessageDigest::sha512(), question.as_bytes())?);
    let answer_hash = hex::encode(hash(MessageDigest::sha512(), answer.as_bytes())?);

    // 构建请求数据
    let data = json!({
        "name": username,
        "newpassword": newpassword_hash,
        "question": question_hash,
        "answer": answer_hash,
        "timestamp": chrono::Utc::now().timestamp_millis(),
    });

    // 修改加密部分
    let encrypted_data = {
        let aeskey_guard = NETDATA.aeskey.lock().unwrap();
        encrypt_json(&data, &aeskey_guard)
    }.map_err(|e| anyhow::anyhow!("加密失败: {}", e))?;

    // 发送修改请求
    let client = reqwest::Client::new();
    let response = client
        .post(&format!("{}/api/v1/users/change", get_host()))
        .header("Content-Type", "application/json")
        .body(json!({
            "data": encrypted_data.data,
            "tag": encrypted_data.tag,
            "iv": encrypted_data.iv
        }).to_string())
        .send()
        .await
        .context("密码修改请求发送失败")?;

    if !response.status().is_success() {
        let error_data: Value = response.json().await?;
        anyhow::bail!(
            "请求失败: {} (原因: {})",
            error_data["error"].as_str().unwrap_or("未知错误"),
            error_data["reason"].as_str().unwrap_or("未知原因")
        );
    }

    // 处理响应
    let response_data: Value = response.json().await?;
    let ciphertext = response_data["data"].as_str().context("无效响应格式")?;
    let iv = response_data["iv"].as_str().context("无效响应格式")?;
    let tag = response_data["tag"].as_str().context("无效响应格式")?;

    // 修改解密部分
    let de_data = {
        let aeskey = NETDATA.aeskey.lock().unwrap();
        decrypt_data(ciphertext, &aeskey, iv, tag)
    }.map_err(|e| anyhow::anyhow!("响应解密失败: {}", e))?;

    // 更新密钥
    if let Some(new_key) = de_data["newkey"].as_str() {
        let decoded_key = general_purpose::STANDARD.decode(new_key)?;
        *NETDATA.aeskey.lock().unwrap() = decoded_key;
    }

    println!("修改密码成功: {:?}", de_data);
    de_data["success"]
        .as_bool()
        .context("无效的成功状态标识")
}